src/backend_model_instance.cc

```diff
@@ -534,27 +536,58 @@ TritonModelInstance::GenerateWarmupData()
   return Status::Success;
 }
 
-void
+Status
+TritonModelInstance::PrepareRequestsForExecution(
+    std::vector<std::unique_ptr<InferenceRequest>>& requests)
+{
+  for (auto& r : requests) {
+    // Load the input states for the inference request.
+    RETURN_IF_ERROR(r->LoadInputStates());
+    // Set request state to signify that request is no longer pending.
+    RETURN_IF_ERROR(r->SetState(InferenceRequest::State::EXECUTING));
+  }
+
+  return Status::Success;
+}
+
+Status
+TritonModelInstance::PrepareRequestsOrRespond(
+    std::vector<std::unique_ptr<InferenceRequest>>& requests)
+{
+  auto status = PrepareRequestsForExecution(requests);
+  // If any errors occurred, respond with error for each request.
+  if (!status.IsOk()) {
+    for (auto& r : requests) {
+      InferenceRequest::RespondIfError(r, status, true /* release_requests */);
+    }
+    // Log a single error for batch of requests for better visibility
+    LOG_STATUS_ERROR(status, "Requests failed pre-execution checks");
+  }
+
+  return status;
+}
+
+Status
 TritonModelInstance::Schedule(
     std::vector<std::unique_ptr<InferenceRequest>>&& requests,
     const std::function<void()>& OnCompletion)
 {
+  // Prepare requests for execution, respond to requests if any error occur.
+  RETURN_IF_ERROR(PrepareRequestsOrRespond(requests));
+
   // Use a thread local vector to avoid needing to malloc each
   // time an inference is run.
   thread_local std::vector<TRITONBACKEND_Request*> triton_requests(1024);
   triton_requests.clear();
   for (auto& r : requests) {
-    // Load the input states for the inference request.
-    r->LoadInputStates();
-    // Set request state to signtify that request is no longer pending.
-    r->SetState(InferenceRequest::State::EXECUTING);
     triton_requests.push_back(
         reinterpret_cast<TRITONBACKEND_Request*>(r.release()));
   }
 
   Execute(triton_requests);
-
   OnCompletion();
```

Comment:
If we keep the error checking / early return above, what do you think of defining and using a [ScopedDefer](https://github.com/triton-inference-server/python_backend/blob/6f369ef10312ad3db0aef73df5b1138b8467cf14/src/scoped_defer.cc#L30-L50)() like in Python Backend for this?

```cpp
Status
TritonModelInstance::Schedule(
    std::vector<std::unique_ptr<InferenceRequest>>&& requests,
    const std::function<void()>& OnCompletion)
{
  // Ensure completion callback is always called regardless of error
  auto deferred_callback = []{ OnCompletion(); }
  ScopedDefer _(deferred_callback);

  // Prepare requests for execution, respond to requests if any error occur.
  RETURN_IF_ERROR(PrepareRequestsOrRespond(requests));
 
  ...
  Execute(triton_requests);
  return Status::Success
}

```

Reply:
Actually, the callback is associated with the payload. The expectations for when it's called I don't think is well documented, but assuming it's intended to be always be called after Execute() as it currently is -- wouldn't it make more sense to call the callback closer to Payload code, such as at end of [payload->Execute()](https://github.com/triton-inference-server/core/blob/8807e87027cb8a38643984c32c71a6e43d4a9c23/src/payload.cc#L214)?

Reply:
After discussing with Guan and Tanmay, I added some clarifying comments on the existing `payload->Callback()` usage. It may be an over-optimization at the moment compared to Release callbacks, but Tanmay and I decided not to change it for now. Tanmay may look into cleaning these callbacks up a bit in the future.

As for calling the callback after Execute regardless of error, I moved the callback into `payload->Execute()` to keep the payload-specific callback with the payload code. No other code currently calls `instance->Schedule()` with any callbacks, so we can simplify the instance logic there.